### PR TITLE
Adding Shutdown Message to Worker Nodes

### PR DESCRIPTION
**Pull request by BenVillalobos**
Making debugging easier, one step at a time

### CODE REVIEW

**Comment by Forgind** [comment]
How is this an improvement on [the line immediately before](https://github.com/microsoft/msbuild/blob/master/src/Build/BackEnd/Node/OutOfProcNode.cs#L432)?

**Comment by BenVillalobos** [comment]
The short answer is I didn't notice it, mainly because I never saw that log message or the files it's generating. Interesting that it creates a fresh file only to write that single line.

Looking at this now I think the logic should actually replace the line you highlighted, so instead of creating and writing to a file it just does a `CommunicationsUtilities.Trace` call.

**Comment by Forgind** [comment]
I'll admit that's a little weird. Maybe replace their file write with CommunicationsUtilities.Trace? I'd like to keep the `if (_debugCommunications)` part.

**Comment by BenVillalobos** [comment]
Agreed, the if statement needs to stay and I'll replace the streamwriter logic with the `Trace` call.

File: src/Build/BackEnd/Node/OutOfProcNode.cs [hunk 1]
```diff
@@ -248,7 +248,8 @@ public NodeEngineShutdownReason Run(bool enableReuse, out Exception shutdownExce
         public NodeEngineShutdownReason Run(bool enableReuse, bool lowPriority, out Exception shutdownException)
         {
             // Console.WriteLine("Run called at {0}", DateTime.Now);
-            string pipeName = NamedPipeUtil.GetPipeNameOrPath("MSBuild" + Process.GetCurrentProcess().Id);
+            int currentProcID = Process.GetCurrentProcess().Id;
```
Comment (Forgind) [inline src/Build/BackEnd/Node/OutOfProcNode.cs:251]: May as well revert this part.

File: src/Build/BackEnd/Node/OutOfProcNode.cs [hunk 2]
```diff
@@ -427,10 +428,7 @@ private NodeEngineShutdownReason HandleShutdown(out Exception exception)
         {
             if (_debugCommunications)
             {
-                using (StreamWriter writer = File.CreateText(String.Format(CultureInfo.CurrentCulture, Path.Combine(Path.GetTempPath(), @"MSBuild_NodeShutdown_{0}.txt"), Process.GetCurrentProcess().Id)))
-                {
-                    writer.WriteLine("Node shutting down with reason {0} and exception: {1}", _shutdownReason, _shutdownException);
-                }
+                CommunicationsUtilities.Trace("Shutting down with reason: {1},  and exception: {2}.", Process.GetCurrentProcess().Id, _shutdownReason, _shutdownException);
```
Comment (Forgind) [inline src/Build/BackEnd/Node/OutOfProcNode.cs:431]: Actually, have you looked at `CommunicationsUtilities.Trace`? It does the same thing, just writing to a different spot:
https://github.com/microsoft/msbuild/blob/master/src/Shared/CommunicationsUtilities.cs#L632-L639

File: src/Build/BackEnd/Node/OutOfProcNode.cs [hunk 3]
```diff
@@ -515,6 +513,11 @@ private NodeEngineShutdownReason HandleShutdown(out Exception exception)
                 CleanupCaches();
             }
 
+            if (_debugCommunications)
+            {
+                CommunicationsUtilities.Trace("Shut down complete.", Process.GetCurrentProcess().Id);
```
Comment (Forgind) [inline src/Build/BackEnd/Node/OutOfProcNode.cs:518]: For both of these, I think the order of your arguments is reversed. See [here](https://github.com/microsoft/msbuild/blob/master/src/Shared/CommunicationsUtilities.cs#L604)

File: src/Build/BackEnd/Node/OutOfProcNode.cs (hunk 2 above)
Comment (BenVillalobos) [reply src/Build/BackEnd/Node/OutOfProcNode.cs:431]: I figured as much. At least with `Trace` it'll be consistent with how we expect to debug node communication.

File: src/Build/BackEnd/Node/OutOfProcNode.cs (hunk 3 above)
Comment (BenVillalobos) [reply src/Build/BackEnd/Node/OutOfProcNode.cs:518]: I'm using the overload [above that one](https://github.com/microsoft/msbuild/blob/master/src/Shared/CommunicationsUtilities.cs#L596). I actually mistakenly left the second argument in there. We don't need to pass the current process ID so I'll get rid of that.

File: src/Build/BackEnd/Node/OutOfProcNode.cs [hunk 4]
```diff
@@ -427,10 +427,7 @@ private NodeEngineShutdownReason HandleShutdown(out Exception exception)
         {
             if (_debugCommunications)
```
Comment (rainersigwald) [inline src/Build/BackEnd/Node/OutOfProcNode.cs:428]: Should we preserve this check? Trace does it inside.

File: src/Build/BackEnd/Node/OutOfProcNode.cs [hunk 5]
```diff
@@ -427,10 +427,7 @@ private NodeEngineShutdownReason HandleShutdown(out Exception exception)
...
             {
-                using (StreamWriter writer = File.CreateText(String.Format(CultureInfo.CurrentCulture, Path.Combine(Path.GetTempPath(), @"MSBuild_NodeShutdown_{0}.txt"), Process.GetCurrentProcess().Id)))
-                {
-                    writer.WriteLine("Node shutting down with reason {0} and exception: {1}", _shutdownReason, _shutdownException);
-                }
+                CommunicationsUtilities.Trace("Shutting down with reason: {1},  and exception: {2}.", Process.GetCurrentProcess().Id, _shutdownReason, _shutdownException);
```
Comment (rainersigwald) [inline src/Build/BackEnd/Node/OutOfProcNode.cs:430]:
```suggestion
                CommunicationsUtilities.Trace("Shutting down with reason: {0},  and exception: {1}.", _shutdownReason, _shutdownException);
```

**Pull request merged**